File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
import sys
sys.setrecursionlimit(1000000)

def DFS(maps,x,y,cnt):
    
    cnt = int(maps[x][y])
    
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    
    maps[x][y] = 'X'
    
    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]
        
        if  0<=nx< len(maps) and 0<=ny<len(maps[0]) and maps[nx][ny] != 'X':
            plus = int(maps[nx][ny])
            cnt += DFS(maps,nx,ny,cnt+plus)
            
    
    return cnt   
    

def solution(maps):
    answer = []
    
    realmaps = []
    for i in range(len(maps)) :
        realmaps.append(list(maps[i]))

    
    for i in range(len(realmaps)) :
        for j in range(len(realmaps[0])):
            if realmaps[i][j] != 'X':
                an = DFS(realmaps,i,j,int(realmaps[i][j]))
                if an == 0 :
                    continue
                else:
                    answer.append(an)
    if len(answer) == 0 :
        return [-1]
    else :
        return sorted(answer)

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;

class Solution {
    
    int [] dx = new int [] {0,1,0,-1};
    int [] dy = new int [] {1,0,-1,0};
    boolean [][] visited ;
    int N,M; 
    List <Integer> answerlist = new ArrayList<>();
    String [] globalmap ;
    public int[] solution(String[] maps) {
        N = maps.length;
        M = maps[0].length();
        visited = new boolean[N][M];
        globalmap = maps;
        
        for (int i = 0 ; i<N; i++)
        {
            for (int j = 0 ; j<M ; j++)
            {
                if (!visited[i][j] && globalmap[i].charAt(j) != 'X')
                {
                    int val = dfs(i,j,0);
                    answerlist.add(val);
                }
            }
        }
        
        Collections.sort(answerlist);
        
        if (answerlist.size() == 0)
        {
            return new int [] {-1};
        }
        
        int [] answerarr = new int [answerlist.size()];
        
            for (int i = 0 ; i<answerlist.size(); i++)
            {
                answerarr[i] = answerlist.get(i);
            }
        
        return answerarr;
    }
    
    public int dfs(int x, int y, int cnt)
    {
        visited[x][y] = true;
        cnt += Character.getNumericValue(globalmap[x].charAt(y));
        
        for (int i = 0 ; i<4; i++)
        {
            int nx = x+dx[i];
            int ny = y+dy[i];
            
            if(0<=nx && nx <N && 0<= ny && ny <M )
            {
                if (!visited[nx][ny] && globalmap[nx].charAt(ny) != 'X')
                {
                    //일 하고
                    cnt = dfs(nx,ny,cnt);
                }
            }
        }
        
        return cnt;
    }
}